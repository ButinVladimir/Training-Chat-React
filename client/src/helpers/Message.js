import PropTypes from 'prop-types';

class Message {
  constructor(id, date, from, to, message) {
    this.id = id;
    this.date = date;
    this.from = from;
    this.to = to;
    this.message = message;
  }
}

const messageShape = {
  id: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  from: PropTypes.string.isRequired,
  to: PropTypes.arrayOf(PropTypes.string),
  message: PropTypes.string.isRequired,
};

export default Message;
export { messageShape };
