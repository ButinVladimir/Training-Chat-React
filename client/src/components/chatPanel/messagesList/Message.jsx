import React from 'react';
import PropTypes from 'prop-types';
import { messageShape } from '../../../helpers/Message';

function Message({
  id,
  date,
  from,
  to,
  message,
  onClickUser,
}) {
  const convertedTo = to
    ? to.map(t => (
      <button
        type="button"
        key={t}
        className="user-select to"
        onClick={() => onClickUser(t)}
      >
        {`@${t}`}
      </button>))
    : [];

  return (
    <li className="chat-message" key={id}>
      <span className="date">{`[${date}]`}</span>
      <button
        type="button"
        className="user-select from"
        onClick={() => onClickUser(from)}
      >
        {`${from}: `}
      </button>
      {convertedTo}
      <span className="message">{message}</span>
    </li>
  );
}

Message.propTypes = {
  ...messageShape,
  onClickUser: PropTypes.func.isRequired,
};

export default Message;
