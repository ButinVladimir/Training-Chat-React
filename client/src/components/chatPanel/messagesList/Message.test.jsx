import React from 'react';
import { shallow } from 'enzyme';
import Message from './Message';

describe('Message', () => {
  it('renders without crashing when message has been sent to all and to is null', () => {
    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={null}
        message="testMessage1"
        onClickUser={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('renders without crashing when message has been sent to all and to is not null', () => {
    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={[]}
        message="testMessage1"
        onClickUser={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('renders without crashing when message has been sent to 1 recipient', () => {
    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={['recipient 1']}
        message="testMessage1"
        onClickUser={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('renders without crashing when message has been sent to 2 recipients', () => {
    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={['recipient 1', 'recipient 2']}
        message="testMessage1"
        onClickUser={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('renders without crashing when message has been sent to 3 recipients', () => {
    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={['recipient 1', 'recipient 2', 'recipient 3']}
        message="testMessage1"
        onClickUser={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it('handles click on from user correctly', () => {
    const mock = jest.fn();

    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={[]}
        message="testMessage1"
        onClickUser={mock}
      />,
    );

    component.find('button.from').simulate('click');

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('fromUser');
  });

  it('handles click on to user correctly', () => {
    const mock = jest.fn();

    const component = shallow(
      <Message
        id="1"
        date="8/28/2018"
        from="fromUser"
        to={['toUser1', 'toUser2']}
        message="testMessage1"
        onClickUser={mock}
      />,
    );

    component.find('button.to').forEach(w => w.simulate('click'));

    expect(mock).toHaveBeenCalledTimes(2);
    expect(mock).toHaveBeenCalledWith('toUser1');
    expect(mock).toHaveBeenCalledWith('toUser2');
  });
});
