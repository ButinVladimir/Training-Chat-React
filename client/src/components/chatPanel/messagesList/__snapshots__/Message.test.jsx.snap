// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message renders without crashing when message has been sent to 1 recipient 1`] = `
<li
  className="chat-message"
  key="1"
>
  <span
    className="date"
  >
    [8/28/2018]
  </span>
  <button
    className="user-select from"
    onClick={[Function]}
    type="button"
  >
    fromUser: 
  </button>
  <button
    className="user-select to"
    key="recipient 1"
    onClick={[Function]}
    type="button"
  >
    @recipient 1
  </button>
  <span
    className="message"
  >
    testMessage1
  </span>
</li>
`;

exports[`Message renders without crashing when message has been sent to 2 recipients 1`] = `
<li
  className="chat-message"
  key="1"
>
  <span
    className="date"
  >
    [8/28/2018]
  </span>
  <button
    className="user-select from"
    onClick={[Function]}
    type="button"
  >
    fromUser: 
  </button>
  <button
    className="user-select to"
    key="recipient 1"
    onClick={[Function]}
    type="button"
  >
    @recipient 1
  </button>
  <button
    className="user-select to"
    key="recipient 2"
    onClick={[Function]}
    type="button"
  >
    @recipient 2
  </button>
  <span
    className="message"
  >
    testMessage1
  </span>
</li>
`;

exports[`Message renders without crashing when message has been sent to 3 recipients 1`] = `
<li
  className="chat-message"
  key="1"
>
  <span
    className="date"
  >
    [8/28/2018]
  </span>
  <button
    className="user-select from"
    onClick={[Function]}
    type="button"
  >
    fromUser: 
  </button>
  <button
    className="user-select to"
    key="recipient 1"
    onClick={[Function]}
    type="button"
  >
    @recipient 1
  </button>
  <button
    className="user-select to"
    key="recipient 2"
    onClick={[Function]}
    type="button"
  >
    @recipient 2
  </button>
  <button
    className="user-select to"
    key="recipient 3"
    onClick={[Function]}
    type="button"
  >
    @recipient 3
  </button>
  <span
    className="message"
  >
    testMessage1
  </span>
</li>
`;

exports[`Message renders without crashing when message has been sent to all and to is not null 1`] = `
<li
  className="chat-message"
  key="1"
>
  <span
    className="date"
  >
    [8/28/2018]
  </span>
  <button
    className="user-select from"
    onClick={[Function]}
    type="button"
  >
    fromUser: 
  </button>
  <span
    className="message"
  >
    testMessage1
  </span>
</li>
`;

exports[`Message renders without crashing when message has been sent to all and to is null 1`] = `
<li
  className="chat-message"
  key="1"
>
  <span
    className="date"
  >
    [8/28/2018]
  </span>
  <button
    className="user-select from"
    onClick={[Function]}
    type="button"
  >
    fromUser: 
  </button>
  <span
    className="message"
  >
    testMessage1
  </span>
</li>
`;
